### PR TITLE
Warn when sequence-level importance sampling is combined with a token-summed loss type

### LINKED ISSUE #3823: Loss not calculated correctly when using GSPO (`importance_sampling_level == "sequence"`)
### Reproduction

```python
from trl import GRPOConfig

training_args = GRPOConfig(
    importance_sampling_level="sequence",
    loss_type="grpo",
    beta=0.00,  
    epsilon=3e-4,
)
```

I believe that currently the GSPO loss is calculated in a different manner than in the paper. If i am misunderstanding something, please let me know.

The loss in the paper is:

Image

which means that each example has only one Advantage.

The current code is:
https://github.com/huggingface/trl/blob/dbbc770f45139b71ecb8a2871b56db1f3ca559d0/trl/trainer/grpo_trainer.py#L1859-L1896

if we use `importance_sampling_level == "sequence"` we get that `per_token_loss` shape is ` (B, 1) `. which makes sense. but when we go to calculate the loss, we multiply `per_token_loss` with `completion_mask` which gives us a shape of ` (B, T) ` (beacuse of broadcasting) which means that the advantage is multiplied by the number of non-padding tokens for each sample. and then we sum/mean to get the loss.

e.g:

Image

as far as i understand this is not how the calculation should be. and when we use GSPO we should just use `per_token_loss.mean() ` as the loss.

I think a similar issue arises if we use kl token loss. which will also have this broadcasting.

Maybe I am misunderstanding something, if that is the case. would love an explanation!

@qgallouedec  

### System Info

- Platform: Linux-5.14.0-503.21.1.el9_5.x86_64-x86_64-with-glibc2.34
- Python version: 3.11.12
- TRL version: 0.19.0.dev0+8f490b3
- PyTorch version: 2.7.0
- accelerator(s): NVIDIA A100-SXM4-80GB
- Transformers version: 4.53.0.dev0
- Accelerate version: 1.7.0
- Accelerate config: not found
- Datasets version: 3.6.0
- HF Hub version: 0.33.4
- bitsandbytes version: 0.46.0
- DeepSpeed version: not installed
- Diffusers version: not installed
- Liger-Kernel version: 0.5.10
- LLM-Blender version: not installed
- OpenAI version: 1.90.0
- PEFT version: 0.15.2
- vLLM version: 0.9.2

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

### PR DESCRIPTION
**Pull request by discobot**
# What does this PR do?

Fixes #3823.

Combining `importance_sampling_level="sequence"` (the GSPO setup) with a loss type that
sums per-token contributions (`bnpo`, `dr_grpo`, `cispo`, or the default `dapo`) silently
length-weights each sequence's loss instead of applying the per-sequence objective GSPO
prescribes.

This PR adds the warning suggested in
https://github.com/huggingface/trl/issues/3823#issuecomment-4539808557, placed right
after the existing `luspo`/`vespo` mismatch checks in `GRPOTrainer.__init__` and
mirroring their style. It points users to `loss_type="grpo"`, matching the GSPO recipe in
the paper index. The experimental `PAPOTrainer` reuses `GRPOTrainer.__init__`, so it
inherits the check.

One thing that turned up while testing: the existing
`test_train_sequence_importance_sampling` itself uses the footgun combo (sequence-level
importance sampling with the default `dapo`), so the new warning now fires during its
setup. I left it unchanged to keep the diff minimal — happy to switch it to
`loss_type="grpo"` if preferred.

Tests: a parametrized test asserting the warning fires for all four affected loss types
(fails on `main` without the fix), and a negative test asserting the recommended GSPO
combo stays silent. Also ran the other warning tests and `test_train_loss_types` locally
(CPU).

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case. (#3823)
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [x] AI-generated: the PR was mostly or fully generated by an AI tool.

The fix and tests were developed with Claude Code and verified by running the test suite locally.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Low Risk**
> Logging-only validation at init; no change to loss computation or training behavior.
> 
> **Overview**
> Adds **init-time warnings** when `importance_sampling_level="sequence"` (GSPO-style) is paired with loss types that **sum per-token terms** (`bnpo`, `dr_grpo`, `dapo`, `cispo`). That combo length-weights each sequence instead of matching the per-sequence GSPO objective; the message recommends `loss_type="grpo"` and links the paper index.
> 
> The check is added in **`GRPOTrainer.__init__`** next to the existing `luspo`/`vespo` config warnings, and the same logic is mirrored in **`SDPOTrainer`**.
> 
> A parametrized test asserts the warning appears for the four affected loss types and stays off for `grpo` with sequence-level IS.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 3fe6e118c4836bf76a2366c6e6a8dbe6306697e2. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Review by qgallouedec — APPROVED** [review]
lgtm!
